kauri/kauri.go

```diff
@@ -0,0 +1,268 @@
+func (k *Kauri) Begin(pc hotstuff.PartialCert, p hotstuff.ProposeMsg) {
+	if !k.initDone {
+		k.eventLoop.DelayUntil(backend.ConnectedEvent{}, func() { k.Begin(pc, p) })
+		return
+	}
+	k.reset()
+	k.blockHash = pc.BlockHash()
+	k.currentView = p.Block.View()
+	k.aggContrib = pc.Signature()
+	k.SendProposalToChildren(p)
+}
+
+func (k *Kauri) reset() {
+	k.aggContrib = nil
+	k.senders = make([]hotstuff.ID, 0)
+	k.aggSent = false
+}
+
+func (k *Kauri) WaitToAggregate(t time.Duration, view hotstuff.View) {
+	ticker := time.NewTicker(t)
+	<-ticker.C
+	ticker.Stop()
+	if k.currentView != view {
+		return
+	}
+	if !k.aggSent {
+		k.SendContributionToParent()
+		k.reset()
+	}
+}
+
+// SendProposalToChildren sends the proposal to the children
+func (k *Kauri) SendProposalToChildren(p hotstuff.ProposeMsg) {
+	children := k.tree.ReplicaChildren()
+	if len(children) != 0 {
+		config, err := k.configuration.SubConfig(children)
+		if err != nil {
+			k.logger.Error("Unable to send the proposal to children", err)
+			return
+		}
+		k.logger.Debug("sending proposal to children ", children)
+		config.Propose(p)
+		waitTime := time.Duration(k.tree.TreeHeight()) * k.treeConfig.TreeWaitDelta()
+		go k.WaitToAggregate(waitTime, k.currentView)
+	} else {
+		k.SendContributionToParent()
+		k.aggSent = true
+	}
+}
+
+// OnContributionRecv is invoked upon receiving the vote for aggregation.
+func (k *Kauri) OnContributionRecv(event ContributionRecvEvent) {
+	if k.currentView != hotstuff.View(event.Contribution.View) {
+		return
+	}
+	contribution := event.Contribution
+	k.logger.Debug("processing the contribution from ", contribution.ID)
+	currentSignature := hotstuffpb.QuorumSignatureFromProto(contribution.Signature)
+	_, err := k.mergeWithContribution(currentSignature)
+	if err != nil {
+		k.logger.Debug("Unable to merge the contribution from ", contribution.ID)
+		return
+	}
+	k.senders = append(k.senders, hotstuff.ID(contribution.ID))
+	if isSubSet(k.tree.SubTree(), k.senders) {
+		k.SendContributionToParent()
+		k.aggSent = true
+	}
+}
+
+// SendContributionToParent sends contribution to the parent node.
+func (k *Kauri) SendContributionToParent() {
+	parent, ok := k.tree.Parent()
+	if ok {
+		node, isPresent := k.nodes[parent]
+		if isPresent {
+			node.SendContribution(context.Background(), &kauripb.Contribution{
+				ID:        uint32(k.opts.ID()),
+				Signature: hotstuffpb.QuorumSignatureToProto(k.aggContrib),
+				View:      uint64(k.currentView),
+			})
+		}
+	}
+}
+
+type serviceImpl struct {
+	k *Kauri
+}
+
+func (i serviceImpl) SendContribution(_ gorums.ServerCtx, request *kauripb.Contribution) {
+	i.k.eventLoop.AddEvent(ContributionRecvEvent{Contribution: request})
+}
+
+// ContributionRecvEvent is raised when a contribution is received.
+type ContributionRecvEvent struct {
+	Contribution *kauripb.Contribution
+}
+
+func (k *Kauri) canMergeContributions(a, b hotstuff.QuorumSignature) bool {
```

Comment:
I believe the following logic is a bit clearer:
```go
func (k *Kauri) canMergeContributions(a, b hotstuff.QuorumSignature) bool {
	if a == nil || b == nil {
		k.logger.Info("kauri: cannot merge nil contributions")
		return false
	}
	a.Participants().RangeWhile(func(i hotstuff.ID) bool {
		// cannot merge a and b if both contain a contribution from the same ID.
		return !b.Participants().Contains(i)
	})
	return true
}
```
But I would be more confident in my statement if we had a test for this one. Could you please add a test?